jaxrs-testing/src/main/java/io/airlift/jaxrs/testing/JaxrsTestingHttpProcessor.java

```diff
@@ -0,0 +1,135 @@
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class JaxrsTestingHttpProcessor
+        implements TestingHttpClient.Processor
+{
+    private static final Logger log = Logger.get(JaxrsTestingHttpProcessor.class);
+
+    private final Client client;
+
+    private boolean trace;
+
+    public JaxrsTestingHttpProcessor(URI baseUri, Object... jaxRsSingletons)
+    {
+        Set<Object> jaxRsSingletonsSet = ImmutableSet.copyOf(jaxRsSingletons);
+        Application application = new Application() {
+            @Override
+            public Set<Object> getSingletons()
+            {
+                return jaxRsSingletonsSet;
+            }
+        };
+        TestContainer testContainer = new InMemoryTestContainerFactory()
+                .create(baseUri, DeploymentContext.newInstance(application));
+        ClientConfig clientConfig = testContainer.getClientConfig();
+        this.client = JerseyClientBuilder.createClient(clientConfig);
+    }
+
+    public JaxrsTestingHttpProcessor setTrace(boolean enabled)
+    {
+        this.trace = enabled;
+        return this;
+    }
+
+    @Override
+    public Response handle(Request request)
+            throws Exception
+    {
+        // prepare request to jax-rs resource
+        MultivaluedMap<String, Object> requestHeaders = new MultivaluedHashMap<>();
+        for (Map.Entry<String, String> entry : request.getHeaders().entries()) {
+            requestHeaders.add(entry.getKey(), entry.getValue());
+        }
+        Invocation.Builder invocationBuilder = client.target(request.getUri()).request().headers(requestHeaders);
+        Invocation invocation;
+        if (request.getBodyGenerator() == null) {
+            invocation = invocationBuilder.build(request.getMethod());
+        }
+        else {
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+            request.getBodyGenerator().write(byteArrayOutputStream);
+            byteArrayOutputStream.close();
+            byte[] bytes = byteArrayOutputStream.toByteArray();
+            Entity<byte[]> entity = Entity.entity(bytes, (String) getOnlyElement(requestHeaders.get("Content-Type")));
+            invocation = invocationBuilder.build(request.getMethod(), entity);
+        }
+
+        // issue request, and handle exceptions
+        javax.ws.rs.core.Response result;
+        try {
+            result = invocation.invoke(javax.ws.rs.core.Response.class);
+        }
+        catch (ProcessingException exception) {
+            // Peel out the exception we threw in jax-rs resource implementation
+            // to facilitate testing exceptional paths
+            if (trace) {
+                log.warn(exception.getCause(), "%-8s %s -> Exception", request.getMethod(), request.getUri());
+            }
+            throw (Exception) exception.getCause();
```

Comment:
This can cause `ClassCastException`, since the cause can be anything. It needs to be wrapped if it doesn't extend `Exception`.